### PR TITLE
[npmi] umap

### PR DESCRIPTION
**Pull request by Sparkier**
* Motivation for features / changes
Add data structure and implementation of umap functionality.

* Technical description of changes
Added a util and modified the data structures to support umap.

### CODE REVIEW

File: tensorboard/webapp/dev_assets/BUILD [hunk 1]
```diff
@@ -6,7 +6,7 @@
 # //tensorboard/webapp:index.html, we had to create the dev_assets version in
 # a different package. Once we can get rid of zip asset provider and a way to
 # alias asset, we would not need the enitrely different Bazel package.
-load("//tensorboard/defs:defs.bzl", "tf_dev_js_binary", "tf_js_binary", "tf_ng_module")
+load("//tensorboard/defs:defs.bzl", "tf_dev_js_binary", "tf_ng_module")
```
Comment (stephanwlee) [inline tensorboard/webapp/dev_assets/BUILD:9]: Things here look extraneous and incorrect. Please fix the commit history and remove unnecessary changes.

Comment (Sparkier) [reply tensorboard/webapp/dev_assets/BUILD]: Should be done.

File: tensorboard/webapp/plugins/npmi/util/sort_annotations.ts [hunk 2]
```diff
@@ -15,27 +15,29 @@ limitations under the License.
...
-    (embeddingData[sort.metric] === undefined && similarityBased)
+    ((embeddingData === undefined ||
+      embeddingData.points[sort.metric] === undefined) &&
+      similarityBased)
   ) {
     return result;
   }
   const distanceData: {[annotation: string]: number} = similarityBased
-    ? calculateDistances(result, embeddingData, sort)
+    ? calculateDistances(result, embeddingData!, sort)
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/util/sort_annotations.ts:40]: Hmm with L33, there should not be a case when `embeddingData` is undefined.

File: tensorboard/webapp/plugins/npmi/util/sort_annotations_test.ts [hunk 3]
```diff
@@ -52,15 +52,11 @@ describe('sort annotations utils', () => {
 
   it('sorts annotations by similar embeddings', () => {
     const annotationData: AnnotationDataListing = buildSampleAnnotationData();
+    const embeddingData = buildSampleEmbeddingData();
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/util/sort_annotations_test.ts:55]: Philosophy thing: As a reader of the code, I should be able to read things in the spec and I should be able to have confidence that the code is correct. With the input folded into a utility, I have to flip through files to know that the sortAnnotation is correct which is not ideal. Can you keep below?
```ts
annotation_1: [0.5],	
annotation_2: [-0.2],
annotation_3: [0.1],
```

File: tensorboard/webapp/plugins/npmi/util/umap.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/util/umap.ts:1]: Test is missing in this module.

File: tensorboard/webapp/plugins/npmi/store/npmi_types.ts [hunk 5]
```diff
@@ -105,6 +112,7 @@ export interface NpmiState {
   embeddingsMetric: Metric;
   embeddingsSidebarWidth: number;
   embeddingsSidebarExpanded: boolean;
+  embeddingStatusMessage: string;
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/store/npmi_types.ts:115]: True, there is no selector yet. It will be set and used for displaying the umap process. I removed it for now.

File: tensorboard/webapp/plugins/npmi/util/sort_annotations_test.ts (hunk 3 above)
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/util/sort_annotations_test.ts:55]: I don't think so. The data structure changed from EmbeddingLising to EmbeddingDataSet, which is no initializable like this.

File: tensorboard/webapp/plugins/npmi/store/npmi_reducers.ts [hunk 6]
```diff
@@ -421,6 +434,23 @@ const reducer = createReducer(
...
+  on(
+    actions.embeddingDataSetChanged,
+    (state: NpmiState, {dataSet}): NpmiState => {
+      const newDataSet = new EmbeddingDataSet(dataSet.points, {
+        pointKeys: dataSet.pointKeys,
+        shuffledDataIndices: dataSet.shuffledDataIndices,
+        projections: dataSet.projections,
+        hasUmapRun: dataSet.hasUmapRun,
+        umapRun: dataSet.umapRun,
+      });
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/store/npmi_reducers.ts:447]: I think I did that. Not sure if everything I did is correct by means of your description.

File: tensorboard/webapp/plugins/npmi/util/sort_annotations.ts (hunk 2 above)
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/util/sort_annotations.ts:40]: I don't get it I think. I get an error that this can be undefined if I remove the `!`. Don't know if that is what you mean, though.

File: tensorboard/webapp/plugins/npmi/util/umap.ts (hunk 4 above)
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/util/umap.ts:1]: I added some tests, but I am not sure if this is the way to go with asynchronous functions. Maybe you can give me some feedback on that.

File: tensorboard/webapp/plugins/npmi/util/umap.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+/** Runs UMAP on the data. */
+export async function projectUmap(
+  embeddingData: EmbeddingDataSet,
+  nNeighbors: number,
+  minDist: number,
+  umapIndices: number[],
+  messageCallback: (message: string) => void,
+  datasetCallback: (dataset: EmbeddingDataSet) => void
+) {
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/util/umap.ts:80]: Specify the return type. Currently, it is set as `Promise<void>`. Might want to remove the datasetCallback and return a `Promise< EmbeddingDataSet>` instead.

File: tensorboard/webapp/plugins/npmi/util/umap.ts [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  return {
+    points,
+    pointKeys,
+    shuffledDataIndices: util.shuffle(util.range(pointKeys.length)),
+    projections: {},
+    hasUmapRun: false,
+  };
+}
+
+/** Runs UMAP on the data. */
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/util/umap.ts:70]: Please make a note to the consumer that this modifies the input `embeddingData` for performance reassons.

File: tensorboard/webapp/plugins/npmi/util/umap.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+export interface EmbeddingDataSet {
+  points: EmbeddingListing;
+  pointKeys: string[];
+  shuffledDataIndices: number[];
+  /**
+   * This keeps a list of all current projections so you can easily test to see
+   * if it's been calculated already.
+   */
+  projections: {
+    [projection: string]: boolean;
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/util/umap.ts:50]: If they can be only three things for now, better type it with
```ts
projections: {
  umap: boolean;
  'umap-1': boolean;
  'umap-2': boolean;
},
```

File: tensorboard/webapp/plugins/npmi/util/umap_test.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+          ).toBeTruthy();
+          expect(
+            embeddingDataSet.points[key].projections['umap-1']
+          ).toBeTruthy();
+        }
+      },
+      () => {
+        fail('UMAP not working as expected.');
+      }
+    );
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/util/umap_test.ts:81]: If you are going to use `await`, this tends to read better.

```ts
const embeddingDataSet = await projectUmap(embeddingDataSet, 2, 0.1, [0, 1, 2, 3, 4], () => {});
expect(embeddingDataSet.projections.umap).toBeTrue();
expect(embeddingDataSet.hasUmapRun).toBeTrue();
for (const key of embeddingDataSet.pointKeys) {
  expect(
    embeddingDataSet.points[key].projections['umap-0']
  ).toBeTruthy();
  expect(
    embeddingDataSet.points[key].projections['umap-1']
  ).toBeTruthy();
}

```

**Review by stephanwlee — COMMENTED** [review]
Almost done! I will merge it the moment comment is fixed.

File: tensorboard/webapp/plugins/npmi/util/umap_test.ts [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {buildEmbeddingDataSet, projectUmap} from './umap';
+import {createSampleEmbeddingListing} from '../testing';
+
+fdescribe('umap utils', () => {
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/util/umap_test.ts:18]: Btw, you cannot have fdescribe :)

File: tensorboard/webapp/plugins/npmi/util/umap.ts [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+export interface EmbeddingDataPoint {
+  vector: number[];
+  name: string;
+  sequenceIndex?: number;
+  index: number;
+  projections: {
+    umap?: boolean;
+    'umap-0'?: number;
+    'umap-1'?: number;
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/util/umap.ts:28]: Just wondering, any reason why you are mixing data type in this structure? How about:
```
interface ... {
  ...,
  projections: {
    umap?: Record<number, number>;
  }
}
```

instead?

**Review by stephanwlee — APPROVED** [review]
I am not going to block based on the type question. Please feel free to respond and/or address after merge or in a follow up.
